initial implementation

H-M-Quang-Ngo opened the pull request:
https://review.opendev.org/c/openstack/watcher-specs/+/943873/6/specs/2025.2/approved/host-maintenance-strategy-disable-migration.rst#17

H-M-Quang-Ngo committed:
initial implementation

H-M-Quang-Ngo commented on watcher/applier/actions/stop_instance.py:
```diff
@@ -0,0 +1,178 @@
+        except Exception as exc:
+            LOG.exception(exc)
+            LOG.warning("Post-condition check failed for instance %s: %s",
+                        self.instance_uuid, str(exc))
```
None of other actions has implemented these two methods, most are just having todo comments.
So not really have a reference. I think here it's maybe reasonable to implement them

H-M-Quang-Ngo commented on watcher/applier/actions/stop_instance.py:
```diff
@@ -0,0 +1,178 @@
+    def execute(self):
+        return self.stop_instance()
+
+    def revert_stop_instance(self):
+        """Revert the stop action by trying to start the instance"""
+        nova = nova_helper.NovaHelper(osc=self.osc)
+        LOG.debug("Starting instance %s", self.instance_uuid)
+
+        try:
+            # Use start_instance from nova_helper - this method exists
+            result = nova.start_instance(instance_id=self.instance_uuid)
+            if result:
+                LOG.info(_("Successfully reverted stop action and started instance %(uuid)s"),
+                         {'uuid': self.instance_uuid})
+            else:
+                LOG.error("Failed to start instance %(uuid)s",
+                          {'uuid': self.instance_uuid})
+            return result
+        except Exception as exc:
+            LOG.exception(exc)
+            LOG.critical("Unexpected error occurred while starting "
+                         "instance %(uuid)s: %(error)s",
+                         {'uuid': self.instance_uuid,
+                          'error': str(exc)})
+            raise
+
+    def revert(self):
+        LOG.debug("Reverting stop action for instance %s", self.instance_uuid)
+        return self.revert_stop_instance()
```
Considering for a revert action, is `start` a valid revert for `stop`...? Also we can implement further condition check for the instance state, i.e don't try to start if instance in certain states ...

jneo8 replied:
In some cases, the OpenStack admin may not have permission to start the instance.
However, it's still valuable to raise this topic upstream for feedback on whether we should retain the revert operation. Otherwise, it could be removed.

H-M-Quang-Ngo commented on watcher/decision_engine/strategy/strategies/host_maintenance.py:
```diff
@@ -174,10 +199,33 @@ def instance_migration(self, instance, src_node, des_node=None):
+
+        # Case 1: Both migrations disabled -> only stop instance
+        if disable_live_migration and disable_cold_migration:
+            self.add_action_stop_instance(instance)
+            return
+
+        # Case 2: Handle active instance
         if instance_state_str == element.InstanceState.ACTIVE.value:
-            migration_type = 'live'
+            if disable_live_migration:
+                # Stop active instance first, then cold migrate
+                self.add_action_stop_instance(instance)
+                if not disable_cold_migration:
+                    migration_type = 'cold'
+                else:
+                    return  # Only stop, no migration
+            else:
+                # Live migrate active instance
+                migration_type = 'live'
         else:
-            migration_type = 'cold'
+            # Case 3: Handle non-active instance
+            if disable_cold_migration:
+                # Non-active instance, cold migration disabled, then do nothing
+                return
+            else:
+                migration_type = 'cold'
```
For the logic here, I'm also wondering, is it usual for a case where the system supports live migration, but not cold migration? Like in case user set `disable_cold_migration`, but not set `disable_live_migration`, then this current logic will still allow live migration.

jneo8 replied:
Yes, then live migration will be apply to those active instances. Non-active instances will be left on the node.

H-M-Quang-Ngo commented on watcher/applier/actions/stop_instance.py:
```diff
@@ -0,0 +1,178 @@
+        LOG.debug("Reverting stop action for instance %s", self.instance_uuid)
+        return self.revert_stop_instance()
+
+    def abort(self):
+        """Abort the stop action - not applicable for stop operations"""
+        LOG.warning("Abort operation is not applicable for stop action on "
+                    "instance %s", self.instance_uuid)
+        return False
```
Should abort be implemented? Currently it looks not really necessary or doable to 'abort' a stop action

jneo8 replied:
There is no abort to cancel the "stop" as I understand.

jneo8 commented on setup.cfg:
```diff
@@ -96,6 +96,8 @@ watcher_actions =
     resize = watcher.applier.actions.resize:Resize
     change_node_power_state = watcher.applier.actions.change_node_power_state:ChangeNodePowerState
     volume_migrate = watcher.applier.actions.volume_migration:VolumeMigrate
+    stop_instance = watcher.applier.actions.stop_instance:StopInstance
```
**suggestion:** ```


Follow the general naming.

```
stop = watcher.applier.actions.stop:Stop
```

jneo8 commented on setup.cfg:
```diff
@@ -113,4 +115,4 @@ watcher_cluster_data_model_collectors =
 count =
-quiet-level = 4
\ No newline at end of file
+quiet-level = 4
```
**nitpick:** 

Not sure what is this.

H-M-Quang-Ngo replied:
Ah, I think that's just the default setting from my code editor. If the file doesn't end with a newline, it will automatically add one whenever I saved a file.

jneo8 commented on watcher/applier/actions/stop_instance.py:
```diff
@@ -0,0 +1,178 @@
+    """Stops a server instance
+
+    This action will allow you to stop a server instance on a compute host.
+    This is typically used when migration is not available or desired.
```
**suggestion:** 
It may not be typically in the future. So think line can be remove. Just describe the behavior of this action.

jneo8 commented on watcher/applier/actions/stop_instance.py:
```diff
@@ -0,0 +1,178 @@
+    def instance_uuid(self):
+        return self.resource_id
+
+    def stop_instance(self):
```
**suggestion:** 

Follow the project naming practice:

```python
def stop(self):
```

jneo8 commented on watcher/applier/actions/stop_instance.py:
```diff
@@ -0,0 +1,178 @@
+        try:
+            result = nova.stop_instance(instance_id=self.instance_uuid)
+            if result:
+                LOG.info(_("Successfully stopped instance %(uuid)s"),
```
**nitpick (non-blocking):** 

I didn't see any action using info level LOG in `actions` package. Maybe be change to debug level.

jneo8 commented on watcher/decision_engine/strategy/strategies/host_maintenance.py:
```diff
@@ -164,6 +183,12 @@ def enable_compute_node_if_disabled(self, node):
+            action_type=self.STOP_INSTANCE,
+            resource_id=instance.uuid)
+
     def instance_migration(self, instance, src_node, des_node=None):
```
**nitpick:** 

The naming of the function may need to be updated. Because now may be only "stop" and not "migration" action.

jneo8 commented on watcher/decision_engine/strategy/strategies/host_maintenance.py:
```diff
@@ -174,10 +199,33 @@ def instance_migration(self, instance, src_node, des_node=None):
+                    migration_type = 'cold'
+                else:
+                    return  # Only stop, no migration
+            else:
```
**nitpick:** 

else & return can be refactored here.

jneo8 reviewed (changes requested):
Some nitpick. Overall is nice.

H-M-Quang-Ngo committed:
Apply recommendations

H-M-Quang-Ngo committed:
Add options to disable live and cold migration in maintenance

This change enhances the Host Maintenance strategy by introducing
two new input parameters: `disable_live_migration` and
`disable_cold_migration`. These parameters allow cloud
administrators to control whether live or cold migration should be
considered during host maintenance operations.

If `disable_live_migration` is set, active instances will be stopped
instead of live migrated. If `disable_cold_migration` is not set,
stopped (inactive) instances will be cold migrated. If
`disable_cold_migration` is set, inactive instances will not be cold
migrated. If both are set, only stop actions will be performed on
instances.

The strategy logic and action plan generation have been updated to
reflect these behaviors. A new "stop" action is introduced and
registered, and the weight planner is updated to handle new action.

Documentation for the Host Maintenance strategy is updated to
describe the new parameters and their effects.

Test Plan:
- Unit tests for HostMaintenance strategy with new parameters
- Integration tests for action plan generation with stop action

Implement watcher-specs: #943873

Change-Id: I201b8e5c52e1bc1a74f3886a0e301e3c0fa5d351
